### PR TITLE
[EDU-1658] - Chat room reactions examples

### PR DESCRIPTION
**Pull request by GregHolmes**
This PR contains the two code samples for the new Examples section we are creating. It contains a Typescript and a React version for Room reactions with Chat. Which will be rendered in the code sandbox live examples in the future, along with others.

The README within each directory /examples/chat-room-reactions/javascript and /examples/chat-room-reactions/react contains information on how to get these running.

[Jira epic ticket](https://ably.atlassian.net/browse/EDU-1658)
[Jira ticket](https://ably.atlassian.net/browse/EDU-1541).

### CODE REVIEW

**Review by splindsay-92 — COMMENTED** [review]
Just a couple comments, but looks good!

File: examples/chat-room-reactions/javascript/src/script.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+const room = chatClient.rooms.get('chat-room-reactions', RoomOptionsDefaults);
+
+/** 💡 Add every every room reaction published to the room 💡 */
+room.reactions.subscribe((reaction) => {
+  const reactionsContainer = document.getElementById('reaction-area');
+
+  const reactionElement = document.createElement('span');
+  reactionElement.className = 'reaction';
+  reactionElement.textContent = reaction.type;
+  reactionElement.dataset.createdAt = new Date().toISOString();
```
Comment (splindsay-92) [inline examples/chat-room-reactions/javascript/src/script.ts:20]: Is there a reason you didn't use `reaction.createdAt` ?

Comment (splindsay-92) [reply examples/chat-room-reactions/javascript/src/script.ts:20]: Ahh I see, you're doing a `setInterval` below with `const currentTime = new Date().getTime();`

File: examples/chat-room-reactions/javascript/src/script.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  reactionsContainer.appendChild(reactionElement);
+});
+
+/** 💡 Attach to the room to subscribe to reactions 💡 */
+await room.attach();
+
+setInterval(() => {
```
Comment (splindsay-92) [inline examples/chat-room-reactions/javascript/src/script.ts:27]: What about having this inside the subscribe block? Would need a small rework, but might make things a bit simpler? You can then just use a setTimeout, what do you think? :)

File: examples/chat-room-reactions/react/src/components/Reaction.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  type: string;
+  headers: Record<string, unknown>;
+  metadata: object;
+  clientId: string;
+  createdAt: string;
+  isSelf: boolean;
+}
+
+export const Reaction = () => {
+  const [reactions, setReactions] = useState<ReactionInterface[]>([]);
```
[the remark this answers is not in the document]
Comment (GregHolmes) [reply examples/chat-room-reactions/react/src/components/Reaction.tsx:16]: I've updated this, but I had to alias Reaction imported as ReactionInterface. Thank you!

File: examples/chat-room-reactions/javascript/src/script.ts (hunk 2 above)
Comment (GregHolmes) [reply examples/chat-room-reactions/javascript/src/script.ts]: Updated!

File: examples/chat-room-reactions/react/src/components/Reaction.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import React, { useState } from 'react';
+import { useRoomReactions } from '@ably/chat/react';
+import { Reaction as ReactionInterface } from '@ably/chat';
+import '../../styles/styles.css'
+
+// interface ReactionInterface {
```
Comment (splindsay-92) [inline examples/chat-room-reactions/react/src/components/Reaction.tsx:6]: Perhaps this was forgotten?

File: examples/chat-room-reactions/react/src/components/Reaction.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import React, { useState } from 'react';
+import { useRoomReactions } from '@ably/chat/react';
+import { Reaction as ReactionInterface } from '@ably/chat';
```
Comment (splindsay-92) [inline examples/chat-room-reactions/react/src/components/Reaction.tsx:3]: What is `ReactionInterface`? I don't think this is a type belonging to chat

File: examples/chat-room-reactions/react/src/components/Reaction.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+// }
+
+export const Reaction = () => {
+  const [reactions, setReactions] = useState<ReactionInterface[]>([]);
+
+  const emojis = ['❤️', '😲', '👍', '😊'];
+
+  const { send } = useRoomReactions({
+    listener: (reaction) => {
+      setReactions((prevReactions: ReactionInterface[]) => [...prevReactions, {...reaction}])
```
Comment (splindsay-92) [inline examples/chat-room-reactions/react/src/components/Reaction.tsx:22]:
```tsx
setReactions((prevReactions: ReactionInterface[]) => [...prevReactions, reaction])
```
Could we just do this? :)

File: examples/chat-room-reactions/javascript/index.html [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        >
+        </div>
+        <div class="border-t-2 border-gray-200 px-4 pt-4 mb-2 sm:mb-0">
+          <form
+            class="flex"
+          >
+            <input
+               type="text"
+              disabled
+              placeholder="Say something"
```
Comment (m-hulbert) [inline examples/chat-room-reactions/javascript/index.html:25]: Should we have something along the lines of `Type a message...` as the placeholder (and add italics?)

File: examples/chat-room-reactions/javascript/index.html [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+              class="w-full focus:outline-none focus:placeholder-gray-400 text-gray-600 placeholder-gray-600 pl-2 pr-2 bg-gray-200 rounded-l-md py-1"
```
Comment (m-hulbert) [inline examples/chat-room-reactions/javascript/index.html:26]: Should we add `disabled:cursor-not-allowed` to the input field too?

File: examples/chat-room-reactions/page.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
+# Room reactions for chat applications
+
+Use room reactions to enable users to express their feelings within a chat application.
+
+Room reactions enable you to render when another chat user send a reaction to the room. They are most commonly used to display emojis such as a thumbs up (👍) or a heart (❤️).
+
+Room reactions can add value to many different chat applications. One example is in 1:1 customer support applications, where agents and clients can express their emotions or feedback instantly using reactions. In larger chat applications, such as a live streaming application, room reactions can enhance user engagement by allowing viewers to react to content in real-time.
```
Comment (m-hulbert) [inline examples/chat-room-reactions/page.md:7]:
```suggestion
Room reactions add value to different applications. They enable users to quickly express their feelings about what is happening on screen. One such example is in a live streaming application where users can use the 👍 emoji if they're enjoying the content, and watch others react in realtime too.
```

File: examples/chat-room-reactions/page.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+Room reactions can be implemented using [Ably Chat](https://ably.com/docs/products/chat). The Chat SDK contains a set of purpose-built APIs that abstract away the complexities involved in architecting chat features. It is built on top of Ably's core platform, and so it provides the same performance guarantees and scaling potential.
```
Comment (m-hulbert) [inline examples/chat-room-reactions/page.md:9]:
```suggestion
Room reactions are implemented using [Ably Chat](https://ably.com/docs/products/chat). The Chat SDK contains a set of purpose-built APIs that abstract away the complexities involved in architecting chat features. It is built on top of Ably's core platform, and so it provides the same performance guarantees and scaling potential.
```

File: examples/chat-room-reactions/page.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+## Resources
+
+// React brief (Only visible if viewing the React example)
+
+Use the following components to add room reactions into a chat application:
+
+* [ChatClientProvider](https://ably.com/docs/chat/setup?lang=react#instantiate): initializes and manages a shared chat client instance, passing it down through React context to enable real-time chat functionality across the application.
+* [ChatRoomProvider](https://ably.com/docs/chat/rooms?lang=react#create): manages the state and functionality of a specific chat room, providing access to messages, participants, and real-time interactions within that room via React context.
+* [useRoom()](https://ably.com/docs/chat/rooms?lang=react#create) hook: a hook to manage the state and interaction for a chat “room”, allowing users to join, send messages, listen for messages, and use the other chat functions such as typing indicators.
+* [useRoomReactions()](https://ably.com/docs/chat/rooms/reactions?lang=react#subscribe) hook: a hook to manage and track the reactions sent by users within a chat room.
```
Comment (m-hulbert) [inline examples/chat-room-reactions/page.md:20]:
```suggestion
* [useRoomReactions()](https://ably.com/docs/chat/rooms/reactions?lang=react#subscribe) hook: a hook to manage the reactions sent by users within a chat room.
```

(careful here, as tracking suggests something like aggregation which isn't supported yet)

**Review by m-hulbert — CHANGES_REQUESTED** [review]
My 2 comments apply to both React and JS 🙂 

A few suggestions to update the text, and then there's an alignment issue (React only) where the emojis are overlapping the 'send' button. This is also interfering with the disabled icon for the same button:

Screenshot 2024-10-21 at 17 14 52

File: examples/chat-room-reactions/javascript/index.html [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+              <div id="reaction-area" className="reaction-area"></div>
+            </div>
+          </div>
+          <form
+            class="flex"
+          >
+            <input
+               type="text"
+              disabled
+              placeholder="Type something..."
```
Comment (m-hulbert) [inline examples/chat-room-reactions/javascript/index.html:31]: I think we used `Start typing...` in typing indicators

File: examples/chat-room-reactions/page.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+## Resources
+
+// React brief (Only visible if viewing the React example)
+
+Use the following components to add room reactions into a chat application:
+
+* [ChatClientProvider](https://ably.com/docs/chat/setup?lang=react#instantiate): initializes and manages a shared chat client instance, passing it down through React context to enable real-time chat functionality across the application.
+* [ChatRoomProvider](https://ably.com/docs/chat/rooms?lang=react#create): manages the state and functionality of a specific chat room, providing access to messages, participants, and real-time interactions within that room via React context.
+* [useRoom()](https://ably.com/docs/chat/rooms?lang=react#create) hook: a hook to manage the state and interaction for a chat “room”, allowing users to join, send messages, listen for messages, and use the other chat functions such as typing indicators.
+* [useRoomReactions()](https://ably.com/docs/chat/rooms/reactions?lang=react#subscribe) hook: a hook to manage the reactions sent by users within a chat room.
```
Comment (m-hulbert) [inline examples/chat-room-reactions/page.md:20]: Should these be code styled like the methods are?

**Review by m-hulbert — COMMENTED** [review]
Functionality all good, just 2 minor formatting comments 🙂

**Review by m-hulbert — APPROVED** [review]
🚀

**Pull request merged**